# 2_call_mutations_mutect2

## 1.Overview  

<img src="https://user-images.githubusercontent.com/64182733/138386177-89a49274-a996-423b-9770-5f6321a959cf.png" width=50%>
  
Mutect2 is a tool for calling short mutations including single nucleotide and insertion and deletion alterations. The caller uses a Bayesian somatic genotyping model and uses the assembly-based machinery of HaplotypeCaller.  

Before calling mutations, a panel of normals need to be established to detect technical artifacts. After calling, there are several filtering steps to exclude single strand lesion and cross sample contamination.

Mutect2 does not support multi-thread. If you have a large number of samples to analysis, refer to Parallelization part to speed up.  
***

## 2.Input Files and Structure
Files|Description
---|--------------
test_tumor.bam|Tumor sample bam file generated by BWA and filtered by GATK toolkit 
test_normal.bam|Normal sample bam file generated by BWA and filtered by GATK toolkit. Optional if you want to run tumor-only mode.
reference.fasta|Refernce genome, take care of version like hg38 or hg19
germline_resource.vcf.gz|Files containing population SNP, can be downloaded from GnomAD, dbSNP or other database. Optional if you cannot find any but recommend.  
test.bed or chr.list|Containing regions to be called, you can specify some regions in .bed or chromosomes in .list. Optional if you to call mutations from the whole genome.   

```
workdir
|---sample_list.txt
|---2.0_patlist.r
|---2.1_pon_1.sh
|---2.1_pon_2.sh
|---2.2_mutect2.sh
|---2.3_readorientation.sh
|---2.4_getpile.sh
|---2.5_calculatecontamination.sh
|---2.6_filter.sh
|---2.7_funcotator.sh
|---data
|      |---test_1_tumor.bam
|      |---test_1_normal.bam
|      |---test_2_tumor.bam
|      |___test_2_normal.bam
|---ref 
|      |___reference.fasta
|---resources
       |---germline_resource.vcf.gz
       |___test.bed or chr.list
```

***

## 3.Create Panel of Normals  

Panel of normals(PON) is made from normal samples (in this context, "normal" means derived from healthy tissue that is believed to not have any somatic alterations).The main purpose is to capture recurrent technical artifacts such sequencing machine artifact or FFPE specific mutations in order to improve the results of the variant calling analysis. 

So it is better to create PON with 40 or more samples from the same extracting and sequencing pipeline. You can also use public PON which can be downloaded from GATK resource bundle.  

* 1.Run mutect2 on normal samples in tumor-only mode
```shell
gatk Mutect2 -R reference.fasta \
  -L test.bed \
  -I test_normal.bam \
  -max-mnp-distance 0 \
  -O normal.vcf.gz
```  

`-max-mnp-distance` is set to 0 to avoid bugs.  

* 2.Create GenomicsDB
```shell
gatk GenomicsDBImport --merge-input-intervals true \
  --genomicsdb-workspace-path pon_db \
  -L test.bed \
  -R reference.fasta \
  -V normal_1.vcf \
  -V normal_2.vcf \
  -V normal_3.vcf
```  

When bed file contain many discrete regions, `GenomicsDBImport` will create to many files. To avoid jam cluster, set `--merge-input-intervals true`.  

* 3.Create panel of normals  
```{r eval=FALSE}
gatk CreateSomaticPanelOfNormals -R reference.fasta \
  -V gendb://pon_db \
  -O pon.vcf.gz
```  
***  
   
## 4.Run mutect2  

Run `Mutect2` to call somatic mutations.

```
gatk Mutect2 -R reference.fasta \
  -I tumor.bam \
  -I normal.bam \
  -L test.bed \
  -normal normal \
  --germline-resource germline_resource.vcf.gz \
  --panel-of-normals pon.vcf.gz \
  --f1r2-tar-gz test_f1r2.tar.gz \
  -O test.vcf.gz
``` 

`-normal` specify the sample name set in `SM` column of bam file header  `@RG\\tID:test\\tSM:normal\\tPL:Illumina`.         
  
***  
   
## 5.Read Orientation Artifacts  

`LearnReadOrientationModel` will recognize mutations which are only on one of two DNA strands.  

```
gatk LearnReadOrientationModel -I test_f1r2.tar.gz \
-O test_orimodel.tar.gz
```  
***  
   
## 6.Tabulate Pileup Metrics and Calculate Contamination  

`GetPileupSummaries` summarizes counts of reads that support reference, alternate and other alleles for given sites. Results can be used with `CalculateContamination`.  

`CalculateContamination` calculates the fraction of reads coming from cross-sample contamination.  

```{r eval=FALSE}
gatk GetPileupSummaries \
  -I normal.bam \
  -V germline_resource.vcf.gz \
  -L test.bed \
  -O normal.table
gatk GetPileupSummaries \
  -I tumor.bam \
  -V germline_resource.vcf.gz \
  -L test.bed \
  -O tumor.table
gatk CalculateContamination \
  -I tumor.table \
  -matched normal.table \
  -O test_contamination.table \
  --tumor-segmentation test_segment.table
```  

***  
   
## 7.Filter Mutations

Filter for confident somatic calls.

```{r eval=FALSE}
gatk FilterMutectCalls -R reference.fasta \
-V test.vcf.gz \
--contamination-table test_contamination.table \
--tumor-segmentation test_segment.table \
--ob-priors test_orimodel.tar.gz \
-O test_filter.vcf.gz
```  

***  
   
## 8.Annotation

Funcotator (FUNCtional annOTATOR) analyzes given variants for their function (as retrieved from a set of data sources) and produces the analysis in a specified output file(like MAF for MutsigCV).

```{r eval=FALSE}
##Download data source
gatk FuncotatorDataSourceDownloader --somatic --validate-integrity --extract-after-download
##Run Funcotator
gatk Funcotator \
     --variant test_filter.vcf.gz \
     --reference Homo_sapiens_assembly38.fasta \
     --ref-version hg38 \
     --data-sources-path funcotator_dataSources.v1.2.20180329 \ #your path to data source downloaded in previous step
     --output test.funcotated.maf \ #remove non-PASS mutations
     --output-file-format MAF \
     --remove-filtered-variants \
     --annotation-default Tumor_Sample_Barcode:test_patient
```  
***  
  
## Optional:Parallelization

For now, mutect2 only support run in single thread. If you want to speed up, here is a way to run mutect2 in multi-threads.  

You can specify `-L` with a chromosome or a .bed file to run mutect2 on a small region, and submit multiple mutect2 jobs at the same time.  

The scripts are stored in `paralleliztion` directory.

```{r eval=FALSE}
gatk Mutect2 -R reference.fasta \
  -L chr1 \ ##chr1 can be substituted by chr2 chr3 ... chr22 chrX chrY
  -I tumor.bam \
  -I normal.bam \
  -normal normal \
  --germline-resource germline_resource.vcf.gz \
  --panel-of-normals pon.vcf.gz \
  --f1r2-tar-gz test_f1r2_chr1.tar.gz \
  -O test_chr1.vcf.gz
```

Then merge results.

```{r eval=FALSE}
gatk MergeVcfs \
  I=test_chr1.vcf.gz \
  I=test_chr2.vcf.gz \
  I=test_chr3.vcf.gz \
  O=test.vcf.gz
gatk MergeMutectStats \
  -stats test_chr1.vcf.gz.stats \
  -stats test_chr2.vcf.gz.stats \
  -stats test_chr3.vcf.gz.stats \
  -O test.vcf.gz.stats
gatk LearnReadOrientationModel \
  -I test_f1r2_chr1.tar.gz \
  -I test_f1r2_chr2.tar.gz \
  -I test_f1r2_chr3.tar.gz \
  -O test_orimodel.tar.gz
```  

Other procedures are the same as mentioned above.

***

## Reference 

https://gatk.broadinstitute.org/hc/en-us
